module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
+	txg_handle_t th;
+	dnode_t *dn;
+	spa_t *spa;
+	uint64_t open_txg, syncing_txg, dirty_txg;
+	int error;
+
+	ZFS_ENTER(zfsvfs);
+	ZFS_VERIFY_ZP(zp);
+
+	error = dnode_hold(zfsvfs->z_os, zp->z_id, FTAG, &dn);
+	if (error) {
+		ZFS_EXIT(zfsvfs);
+		return (error);
+	}
+
+	spa = dmu_objset_spa(dn->dn_objset);
+
+	if (fm->fm_flags & FIEMAP_FLAG_SYNC)
+		txg_wait_synced(spa_get_dsl(spa), 0);
+
+	/*
+	 * Lock the entire file against changes while assembling the FIEMAP.
+	 * Then hold open the TXG while generating a map of all pending frees
+	 * and dirty blocks.  This isn't strictly necessary but it is a
+	 * convenient way to determine the range of TXGs to check.
+	 */
+	rl_t *rl = zfs_range_lock(&zp->z_range_lock, 0, UINT64_MAX, RL_READER);
+	open_txg = txg_hold_open(spa_get_dsl(spa), &th);
+	syncing_txg = dirty_txg = spa_syncing_txg(spa);
+
+	(void) dbuf_generate_dirty_maps(dn, fm->fm_dirty_tree,
+	    fm->fm_free_tree, &dirty_txg, open_txg);
+
+	/*
+	 * When the currently syncing TXG could not be checked, likely because
+	 * the dnode was already synced, we need to wait for the syncing TXG
+	 * to fully complete in order to avoid using stale block pointers.
+	 */
+	if (dirty_txg > syncing_txg)
+		txg_wait_synced(spa_get_dsl(spa), syncing_txg);
+
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	mutex_enter(&dn->dn_mtx);
+
+	dnode_phys_t *dnp = dn->dn_phys;
+	fm->fm_file_size = i_size_read(ip);
+	fm->fm_block_size = dnp->dn_datablkszsec << SPA_MINBLOCKSHIFT;
+	fm->fm_fill_count = 0;
+
+	/*
+	 * When there are only pending dirty buffers the block size will not
+	 * yet have been determined.  Assume the maximum block size.
+	 */
+	if (fm->fm_block_size == 0)
+		fm->fm_block_size = zfsvfs->z_max_blksz;
+
+	/*
+	 * When FIEMAP_FLAG_NOMERGE is set and the number of extents for the
+	 * entire file are being requested.  Then in this special case walking
+	 * the entire indirect block tree is not required.
+	 *
+	 * The fill count can be used to determine the number of extents, all
+	 * dirty blocks are assumed to fill holes, and free blocks are assumed
+	 * to overlap with existing free blocks.  This is a safe worst case
+	 * estimate which may slightly over report the number of extents for
+	 * a file being actively overwritten.
+	 *
+	 * Otherwise, the entire block tree needs to be walked to determine
+	 * exactly how the block pointer will be merged.
+	 */
+	if (fm->fm_extents_max == 0 && fm->fm_flags & FIEMAP_FLAG_NOMERGE &&
+	    fm->fm_start == 0 && fm->fm_length == FIEMAP_MAX_OFFSET) {
+		for (int i = 0; i < MIN(dnp->dn_nblkptr, fm->fm_copies); i++)
+			fm->fm_fill_count += BP_GET_FILL(&dnp->dn_blkptr[i]);
+
+		fm->fm_fill_count += P2ROUNDUP(range_tree_space(
+		    fm->fm_dirty_tree), fm->fm_block_size) / fm->fm_block_size;
+	} else {
+		SET_BOOKMARK(&czb, dmu_objset_id(dn->dn_objset),
+		    dn->dn_object, dnp->dn_nlevels - 1, 0);
+
+		for (int i = 0; i < MIN(dnp->dn_nblkptr, fm->fm_copies); i++) {
+			blkptr_t *bp = &dnp->dn_blkptr[i];
```

Comment:
Could this be running during spa_sync()?  If so, we need the proper locks to prevent the dnode_phys from changing (specifically it's BP's).  See dmu_buf_lock_parent().

Reply:
Right now this is being protected by the `dn_mtx` which should prevent `spa_sync()` from changing the `dnode_phys`.  Which is pretty heavy handed and goes to your larger point of this potentially taking a long time.  I'll see what can be done tp chunk this up so never end up stalling the entire pool.